### DISCUSSION TITLE
all in one (audiokit, faust, Encoder, espnow) ... too much ?

### DISCUSSION THREAD

**Discussion by mkoenings** [pschatzmann/arduino-audio-tools#337] [Q&A] [answered]
Hello, 
i created a Frankenstein ino out of the faust guitarix and the encoded espnow examples.
It compiles, but it does not work, the  serial monitor captures 
_[W] Communication.h : 223 - Write failed - retrying again_
like every 2 seconds.
Where can I look ?  Thank you!

`/**
 * @file streams-i2s-faust_guitarix-i2s.ino
 * @author Phil Schatzmann
 * @brief Example how to use Faust when Faust expects input and provides kitput
 * @version 0.1
 * @date 2022-04-22
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioFaust.h"
#include "guitarix.h"
#include "AudioLibs/Communication.h"
#include "AudioCodecs/CodecSBC.h"

AudioKitStream kit; 

ESPNowStream now;
EncodedAudioStream encoder(&now, new SBCEncoder()); // encode and write to ESP-now

const char *peers[] = {"A8:48:FA:0B:93:01"};

FaustStream<mydsp> faust(kit); // final output of Faust is encoder
StreamCopy copier1(faust, kit);  // copy data from kit to faust
StreamCopy copier2(encoder, faust);  // copies sound from faust to espnow

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // Setup Faust
  auto cfg = faust.defaultConfig();
  faust.begin(cfg);

  // Tube Screemer
  faust.setLabelValue("drive", 0.23);
  faust.setLabelValue("level", -12.6);
  faust.setLabelValue("tone", 489);
  // preamp
  faust.setLabelValue("Pregain", 0.4);
  faust.setLabelValue("Gain", 0.4);
  //jcm 2000
  faust.setLabelValue("Treble", 0.9);
  faust.setLabelValue("Middle", 0.76);
  faust.setLabelValue("Bass", 0.62);
  //Cabinet
  faust.setLabelValue("amount", 100);

  // start I2S
  auto cfg_i2s = kit.defaultConfig(RXTX_MODE); // RX_MODE should be suficcient?
  cfg_i2s.sample_rate = cfg.sample_rate; 
  cfg_i2s.channels = cfg.channels;
  cfg_i2s.bits_per_sample = cfg.bits_per_sample;
  cfg_i2s.input_device = AUDIO_HAL_ADC_INPUT_LINE1;

  
 
  // start encoder
  
  kit.begin(cfg_i2s);
  //encoder.begin(cfg);
  auto cfg2 = now.defaultConfig();
  cfg2.mac_address = "A8:48:FA:0B:93:02";
  now.begin(cfg2);
  now.addPeers(peers);
  
  
}

// Arduino loop - copy sound to kit 
void loop() {

  copier1.copy(); //kit to faust
  copier2.copy(); //faust to encoder
  
}
`

**Comment by mkoenings** [comment]
I just wanted to emphasize: if this is possible, all on one MCU, with this few lines of code, that would be more than awesome! :-)

**Comment by mkoenings** [comment]
yes, i forgot to un-comment _encoder.begin(cfg);_
Now it crashes.... Any Ideas ?
Thanks!

**Comment by pschatzmann** [comment]
My recommendation would be to build complex sketches step by step. 
Otherwise you have no chance to figure out what went wrong!

I am not 100% sure what you try to do but but I guess the processing chain should be something like the following
```
                                                            -> AudioKit
FaustStream - Copy->  MultiOutput
                                                           -> EncodedAudioStream -> ESPNowStream
```

Not sure if this will fit into the ESP32. If you don't get a stable signal you might need to decrease the sampling rate...

**Comment by mkoenings** [comment]
Hi Phil, thanks for looking into this.
In simple words: I want to transmit processed audio data from a guitar
(line in)  wirelessly in good quality to a receiver.

so the chain would be:
AudioKit -> Faust (as effect processor)  -> Encoder -> ESPNow?
This is what I tried to achieve in the code. Does that make sense ?


Am Mo., 5. Sept. 2022 um 15:33 Uhr schrieb Phil Schatzmann <
***@***.***>:

**Comment by pschatzmann** [comment]
AudioKit -> Faust (as effect processor)  -> Encoder -> ESPNow would give:

```
ESPNowStream now;
EncodedAudioStream encoder(&now, new SBCEncoder()); // encode and write to ESP-now
FaustStream<mydsp> faust(encoder); // final output of Faust is encoder
StreamCopy copier(faust, kit);  // copy data from kit to faust
```

**Comment by mkoenings** [comment]
ah, ok. I got the faust constructor wrong.
And its needs only one copier, thats great.
It doesnt crash anymore but it still fails with

**Comment by mkoenings** [comment]
AudioCopy.h : 256 - write to target has failed!
 in a fast sequence.

**Comment by mkoenings** [comment]
`/**
 * @file streams-i2s-faust_guitarix-i2s.ino
 * @author Phil Schatzmann
 * @brief Example how to use Faust when Faust expects input and provides kitput
 * @version 0.1
 * @date 2022-04-22
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioFaust.h"
#include "guitarix.h"
#include "AudioLibs/Communication.h"
#include "AudioCodecs/CodecSBC.h"



/////



/////

AudioKitStream kit; 



ESPNowStream now;
EncodedAudioStream encoder(&now, new SBCEncoder()); // encode and write to ESP-now

const char *peers[] = {"A8:48:FA:0B:93:01"};




FaustStream<mydsp> faust(encoder); // final output of Faust is encoder
StreamCopy copier(faust, kit);  // copy data from kit to faust


// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // Setup Faust
  auto cfg = faust.defaultConfig();
  faust.begin(cfg);

  // Tube Screemer
  faust.setLabelValue("drive", 0.23);
  faust.setLabelValue("level", -12.6);
  faust.setLabelValue("tone", 489);
  // preamp
  faust.setLabelValue("Pregain", 0.4);
  faust.setLabelValue("Gain", 0.4);
  //jcm 2000
  faust.setLabelValue("Treble", 0.9);
  faust.setLabelValue("Middle", 0.76);
  faust.setLabelValue("Bass", 0.62);
  //Cabinet
  faust.setLabelValue("amount", 100);


  // start I2S
  auto cfg_i2s = kit.defaultConfig(RXTX_MODE); // RX_MODE should be suficcient?
  cfg_i2s.sample_rate = cfg.sample_rate; 
  cfg_i2s.channels = cfg.channels;
  cfg_i2s.bits_per_sample = cfg.bits_per_sample;
  cfg_i2s.input_device = AUDIO_HAL_ADC_INPUT_LINE1;

  
 
  // start encoder
  
  kit.begin(cfg_i2s);
  encoder.begin(cfg);
  auto cfg2 = now.defaultConfig();
  cfg2.mac_address = "A8:48:FA:0B:93:02";
  now.begin(cfg2);
  now.addPeers(peers);
  
  
}

// Arduino loop - copy sound to kit 
void loop() {

  copier.copy();
  
}
`

**Comment by pschatzmann** [comment]
The log might give a clue what is wrong. 
In addition you can play around by replacing some output classes.

FaustStream<mydsp> faust(encoder); // final output of Faust is encoder
with  
FaustStream<mydsp> faust(now); // final output of Faust is encoder

To identify if the issue is with the encoder.

Or replace the final output and use e.g. a HexDumpStream instead ESPNowStream to figure out if the issue is related to ESPNow...

**Comment by mkoenings** [comment]
Thanks, I will keep trying and keep you posted.

Phil Schatzmann ***@***.***> schrieb am Mo., 5. Sept. 2022,
20:23:

**Comment by mkoenings** [comment]
when faust is bypassed, it works fine.
When I write:
  *FaustStream<mydsp> faust(encoder); // final output of Faust is encoder*
I get this in the start up log:
[E] AudioFaust.h : 252 - Faust expects *input* - you need to provide and
AudioStream in the constructor
  which is confusing since the constructor parameter is for the 'out'
stream ?
Thanks...




Am Di., 6. Sept. 2022 um 07:37 Uhr schrieb Michael Könings <
***@***.***>:

**Comment by mkoenings** [comment]
that has not the desired outcome.
If I put *kit* in the constructor of the FaustStream, it asks for RXTX_MODE
( error if only RX is enabled) , and if RXTX_MODE is set, it plays to the
lineout of the kit, not to the encoder and espnow.
So there is a knot somewhere :-)
Thanks, Michael

Am Di., 6. Sept. 2022 um 13:05 Uhr schrieb Phil Schatzmann <
***@***.***>:

**Reply by pschatzmann** [reply]
I deleted my wrong answer from above and confirm. The parameter in the constructor is for the output!
So your version above should be correct!

FaustStream faust(encoder); // final output of Faust is encoder
FaustStream faust(now); // final output of Faust is encoder

**Comment by mkoenings** [comment]
ok, 
_FaustStream faust(encoder); // final output of Faust is encoder_
seems to be what I need, but it does not work. It creates a message at startup:  
[E] AudioFaust.h : 252 - Faust expects **input** - you need to provide an AudioStream in the constructor

So the question is: how can we provide a different input (kit) AND output (encoder) to faust. Faust seems to work with kit only because kit has two 
 stream "ends" for input amd output. ( my naive assumption).
thanks!

**Comment by pschatzmann** [comment]
I will double check tomorrow: but somehow this can't really be: You can double check this in the FaustStream class: this message only comes when p_out is null and p_out gets assigned from the AudioStream that you pass in the constructor. There is also no logic to clear p_out.

**Comment by pschatzmann** [comment, marked as answer]
I adapted the constructor slightly to accept a more generic Print object: Please use the latest version.
After this change I adapted the example to output to CsvStream

```
CsvStream<int16_t> out(Serial, 1);
AudioKitStream kit; 
FaustStream<mydsp> faust(out); // final output of Faust is kit
StreamCopy copier(faust, kit);  // copy data from kit to faust
```

kit.defaultConfig(RX_MODE) is also working fine. 
So, I can't reproduce your issue...

**Comment by mkoenings** [comment]
Hi Phil, it works now! the whole chain: kit->faust->encoder-espnow , and I can receive it with the decoder espnow example on a second kit. This is simply amazing! Thank you very much. I can now create my own faust files ans start playing with that.
I  will also measure the latency ( which "feels" very short)  and the drifting of the latency.
For may project, I am looking for a rock solid latency, with good but not CD quality. I will keep you posted :-)
Thanks and all the best, Michael

**Comment by pschatzmann** [comment]
Cool: I am interested in your results. 
I made this functionality to build some guitar effects, but so far I haven't managed to start this project...
